src/app/gps/qgsgpstoolbar.cpp

```diff
@@ -0,0 +1,255 @@
+    copyright            : (C) 2022 Nyall Dawson
+    email                : nyall dot dawson at gmail dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsgpstoolbar.h"
+#include "qgsappgpsconnection.h"
+#include "qgscoordinatetransform.h"
+#include "qgsmapcanvas.h"
+#include "qgsproject.h"
+#include "qgis.h"
+#include "qgscoordinateutils.h"
+#include "qgisapp.h"
+#include "qgsappgpssettingsmenu.h"
+#include "qgsapplication.h"
+
+#include <QLabel>
+#include <QToolButton>
+
+QgsGpsToolBar::QgsGpsToolBar( QgsAppGpsConnection *connection, QgsMapCanvas *canvas, QWidget *parent )
+  : QToolBar( parent )
+  , mConnection( connection )
+  , mCanvas( canvas )
+{
+  setObjectName( QStringLiteral( "mGpsToolBar" ) );
+  setWindowTitle( tr( "GPS Toolbar" ) );
+  setToolTip( tr( "GPS Toolbar" ) );
+
+  mWgs84CRS = QgsCoordinateReferenceSystem::fromOgcWmsCrs( QStringLiteral( "EPSG:4326" ) );
+
+  mConnectAction = new QAction( tr( "Connect GPS" ), this );
+  mConnectAction->setToolTip( tr( "Connect to GPS" ) );
+  mConnectAction->setIcon( QgsApplication::getThemeIcon( QStringLiteral( "/gpsicons/mIconGpsConnect.svg" ) ) );
+  mConnectAction->setCheckable( true );
+  addAction( mConnectAction );
+
+  connect( mConnectAction, &QAction::toggled, this, [ = ]( bool connect )
+  {
+    if ( connect )
+      mConnection->connectGps();
+    else
+      mConnection->disconnectGps();
+  } );
+
+  mRecenterAction = new QAction( tr( "Recenter" ) );
+  mRecenterAction->setToolTip( tr( "Recenter map on GPS location" ) );
+  mRecenterAction->setIcon( QgsApplication::getThemeIcon( QStringLiteral( "/gpsicons/mActionRecenter.svg" ) ) );
+  mRecenterAction->setEnabled( false );
+
+  connect( mRecenterAction, &QAction::triggered, this, [ = ]
+  {
+    if ( mConnection->lastValidLocation().isEmpty() )
+      return;
+
+    const QgsCoordinateTransform canvasToWgs84Transform = QgsCoordinateTransform( mCanvas->mapSettings().destinationCrs(), mWgs84CRS, QgsProject::instance() );
+    try
+    {
+      const QgsPointXY center = canvasToWgs84Transform.transform( mConnection->lastValidLocation(), Qgis::TransformDirection::Reverse );
+      mCanvas->setCenter( center );
+      mCanvas->refresh();
+    }
+    catch ( QgsCsException & )
+    {
+
+    }
+  } );
+  addAction( mRecenterAction );
+
+  addSeparator();
+
+  mAddTrackPointAction = new QAction( tr( "Add Track Point" ), this );
```

Comment:
As above, consider changing to Add Track Vertex to match nomenclature used elsewhere.